### PR TITLE
Move edits to their own collection

### LINKED ISSUE #4320: UserEdit document can exceed MongoDB's 16 MB limit, permanently blocking goal/step writes
**Describe the bug**

Once a user's `UserEdit` document for a project grows close to MongoDB's hard
16 MB per-document limit, every subsequent goal/step write to it fails with a 500,
and the user is permanently unable to record new work in that project.

Reads still succeed, so the document is not corrupt — the failure is purely on write.

Observed in production (Honeycomb):

- Route: `POST v1/projects/{projectId}/useredits/{userEditId}`
- `http.response.status_code: 500`
- `error.type: MongoDB.Driver.MongoWriteException`
- `exception.message`: `WriteError { Code: 17419, Message: "Resulting document after
  update is larger than 16777216" }`
- Stack: `UserEditRepository.Replace (line 94)` ← `UserEditService.AddGoalToUserEdit
  (line 55)` ← `UserEditController.UpdateUserEditGoal (line 136)`

The affected `UserEdit` had 500+ entries in its `edits` array, sitting just under 16 MB;
the payload to add was only 78 bytes.

**Root cause**

Two design factors combine:

1. **The `edits` array is unbounded** (`UserEdit.Edits`), and each `Edit` carries a
   `changes` JSON blob plus `stepData`. It grows with every goal a user completes and
   is never trimmed.
2. **Writes rewrite the entire document.** `AddGoalToUserEdit`, `AddStepToGoal`, and
   `UpdateStepInGoal` all load the full `UserEdit`, mutate the array in memory, and call
   `UserEditRepository.Replace` → `ReplaceOneAsync` with the whole document. When the
   already-large document plus one more edit crosses 16 MB, Mongo rejects the write.

Because it's a full-document replace, there is no way for the user to make further
progress in that project once the threshold is reached — the array can only grow.

**To Reproduce**

1. As a single user, accumulate a very large number of goal edits in one project
   (or seed a `UserEdit` whose `edits` array approaches 16 MB).
2. Attempt to start/advance another goal (POST to `.../useredits/{userEditId}`) or a
   step (PUT to the same route).
3. Backend returns 500; Honeycomb shows `MongoWriteException` code 17419.

**Expected behavior**

Recording goals/steps should not fail as edit history accumulates. The stored history
should be bounded (or stored such that no single document can hit the 16 MB limit), and
writes should not require rewriting the whole document.

**Proposed fixes** (for discussion)

- Short term: replace the append-then-`Replace` pattern with an atomic `$push` +
  `$slice` that caps the `edits` array length server-side, bounding growth going forward.
- Consider trimming/limiting what is stored in each `Edit.changes`.
- Longer term: move edits to their own collection (one document per edit) so no single
  document is bounded by the 16 MB limit. (Schema migration.)

**Note on already-affected data**

At least one production `UserEdit` is already at the limit and needs its `edits` array
trimmed in place to unblock the user. Trim in place — do not delete the document, since
the user's `WorkedProjects` map references its `_id`.

**Environment**

- Component: Backend (`UserEditController` / `UserEditService` / `UserEditRepository`)
- Server: production (`thecombine.app`)
- Database: MongoDB (16 MB BSON document limit)

### PR DESCRIPTION
**Pull request by imnasnainaec**
Resolves #4320 (the "longer term" fix proposed there). Supersedes #4327.

## Problem

A `UserEdit` document embeds its entire `edits` array, which grows without bound, and every goal/step write loaded the full document and rewrote it with `ReplaceOneAsync`. Once a document neared MongoDB's 16 MB limit, every write failed with `WriteError 17419` and the user was permanently blocked from recording goal/step progress in that project (observed in production).

## Fix

Each `Edit` now lives in its own document in a new `EditsCollection` (`StoredEdit`: `projectId`, `userEditId`, plus the edit fields), and the `UserEdit` document's `edits` array holds only ObjectId refs, preserving order (`StoredUserEdit`). No document now grows meaningfully with use, so there is no cap on edit history.

- Goal/step writes are targeted atomic updates on the small per-edit documents;
  - `UserEditRepository.Replace` (full-document rewrite) is removed.
- Appending a new edit inserts the `StoredEdit` and pushes its ref inside a multi-document transaction (the DB always runs as a replica set), so a failed append can't leave orphans.
- Reads assemble the wire `UserEdit` from both collections, so **the API contract is unchanged**;
  - no frontend or OpenAPI changes.
- `rm_project.py` now also purges `EditsCollection`.

## Mongo migration instructions (required)

`database/migrate-useredits-to-edits-collection.js`:
- moves every embedded edit into `EditsCollection`;
- replaces each `edits` array with ordered refs;
  - documents are modified in place, not deleted, since each user's `workedProjects` references the document `_id`;
- creates the `{ projectId, userEditId, guid }` index;
- verifies that no old-format documents remain, that every ref resolves to exactly one edit document (comparing ids, not counts), that each edit's `projectId` matches its parent's, and that no edit is orphaned or unreferenced — throwing, so a failed verification exits non-zero.

It is idempotent and safe to rerun.

This is a breaking schema change — old backends can't read migrated documents and the new backend can't read unmigrated ones — so it must be coordinated with the deploy:

1. **Back up** (run `combine_backup.py` in the maintenance pod)
2. **Stop the backend** (`kubectl -n thecombine scale deployment backend --replicas=0`)
3. **Deploy the update**
4. **Stop the backend again** (the deploy scales it back to 1)
5. **Run the migration** on the database pod:

   ```bash
   kubectl -n thecombine cp database/migrate-useredits-to-edits-collection.js <database-pod>:/tmp/migrate-useredits-to-edits-collection.js
   kubectl -n thecombine exec <database-pod> -- mongosh CombineDatabase /tmp/migrate-useredits-to-edits-collection.js
   ```

6. **Start the backend** (`kubectl -n thecombine scale deployment backend --replicas=1`)

Per the precedent of the GUID-subtype migration (#4179 / #4209), the script can be removed from the repo in a follow-up once it has been run everywhere.

## Testing

- 20 new `UserEditRepositoryTests` integration tests against a real (ephemeral) mongod replica set
- The repository integration fixtures now share one ephemeral mongod via a `SetUpFixture`; two runner instances in one test process caused intermittent `MongoNotPrimaryException` flakiness.
- Existing `UserEditControllerTests` pass unchanged, confirming the API behavior is untouched.
- The raw BSON layout is asserted in tests (the `edits` refs serialize as ObjectIds), so it matches what the migration script writes.

## Notes for reviewers

- Already-affected production documents are fully unblocked by the migration.
- `AddGoalToUserEdit` now decides add-vs-replace from the atomic `ReplaceEdit` result instead of a prior read, and `UpdateStepInEdit` requires the target step to exist in its filter, so neither the goal write nor an out-of-bounds step write can corrupt data under concurrency.
  - One pre-existing, low-risk race remains and is intentionally out of scope: `UserEditController.UpdateUserEditStep` still reads the step count to choose append-vs-overwrite, so concurrent step writes to the *same* goal could pick the wrong branch. Fixing it would require reshaping the wire contract (the client sends `StepIndex ?? append`), and the atomic bounds guard already rules out the only data-corruption outcome, so this is left for a follow-up.
- The `EditsCollection` index is `{ projectId, userEditId, guid }`, covering both the per-user-edit listing and the single-edit replace/step queries.
  - The backend doesn't manage indexes anywhere today, so new `database/init/01-indexes.js` creates it from the database pod's postStart hook on every container start, alongside `00-replica-set.js`. Fresh installs are therefore covered, and the index outlives the migration script's eventual removal. The migration creates it too, so a just-migrated database is indexed before its next restart.
- `UserEditController.UpdateUserEditStep` now takes the *first* edit with a given guid rather than the last, agreeing with the repository's `UpdateOne` step writes; previously a duplicate guid would have had the step count read from one edit and the write land on another.
- The migration script called `Object.bsonsize`, which mongosh does not provide, so it aborted on its first document with a `TypeError`. Document sizes now come from the server-side `$bsonSize` operator. The script has since been exercised end to end against a real mongod: a clean run, an idempotent rerun, recovery from an interrupted run, and each of five injected corruptions detected with a non-zero exit.

Co-authored by Claude.

Devin: https://app.devin.ai/review/sillsdev/TheCombine/pull/4328

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/TheCombine/4328)

## Summary by CodeRabbit

* **New Features**
  * Added support for adding, replacing, and updating individual goals and steps within user edits.
  * Improved handling of missing edits and invalid step positions with clear operation results.

* **Bug Fixes**
  * Improved project-level data isolation when retrieving and deleting user edits.
  * Ensured project removal also cleans up associated edit data.
  * Improved reliability when migrating and validating existing user-edit data.
  * Corrected updates to target the intended matching edit.

* **Tests**
  * Expanded coverage for user-edit operations, persistence, ordering, isolation, and edge cases.

### CODE REVIEW

File: Backend/Models/UserEdit.cs [hunk 1]
```diff
@@ -87,4 +87,79 @@ public Edit Clone()
...
+    public class StoredUserEdit
+    {
+        [BsonId]
+        [BsonRepresentation(BsonType.ObjectId)]
+        public string Id { get; set; } = "";
+
+        /// <summary> Ids of this user edit's <see cref="StoredEdit"/> documents, in order. </summary>
+        [BsonElement("edits")]
+        [BsonRepresentation(BsonType.ObjectId)]
+        public List<string> EditIds { get; set; } = [];
```
[the remark this answers is not in the document]
Comment (imnasnainaec) [reply Backend/Models/UserEdit.cs:86]: Theoretically true, but not practically a problem.

File: Backend/Services/UserEditService.cs [hunk 2]
```diff
@@ -24,58 +24,31 @@ public UserEditService(IUserEditRepository userEditRepo)
...
-                userEdit.Edits.Add(edit);
-            }
-
-            // Replace the old UserEdit object with the new one that contains the new/updated edit
-            var editReplaced = await _userEditRepo.Replace(projectId, userEditId, userEdit);
+            // Replace the Edit with this guid if present, otherwise add it. Each repo call is atomic on its
+            // own. Two concurrent first-time writes of the same new guid could still both add it, but guids
+            // are client-generated and unique per goal, so that window is tolerated rather than locked.
+            var isSuccess = await _userEditRepo.ReplaceEdit(projectId, userEditId, edit)
+                || await _userEditRepo.AddEdit(projectId, userEditId, edit);
```
[the remark this answers is not in the document]
Comment (imnasnainaec) [reply Backend/Services/UserEditService.cs:40]: Too costly - not worth it. This pr makes an improvement on `master`.

File: database/migrate-useredits-to-edits-collection.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+userEdits.find().forEach(function (doc) {
+  var refCount = doc.edits.length;
+  var docCount = edits.countDocuments({ userEditId: doc._id.toHexString() });
+  if (refCount !== docCount) {
+    failures++;
+    print(
+      "WARNING: UserEdit " + doc._id.toHexString() + " has " + refCount +
+      " ref(s) but " + docCount + " EditsCollection document(s)."
+    );
+  }
```
[the remark this answers is not in the document]
Comment (imnasnainaec) [reply database/migrate-useredits-to-edits-collection.js]: Done.

File: Backend/Controllers/UserEditController.cs [hunk 4]
```diff
@@ -174,8 +174,9 @@ public async Task<IActionResult> UpdateUserEditStep(string projectId, string use
                 return NotFound($"userEditId: {userEditId}");
             }
 
-            // Ensure Edit exist.
-            var edit = document.Edits.FindLast(e => e.Guid == stepWrapper.EditGuid);
+            // Ensure Edit exist. First match, since the repo's step writes are UpdateOne
+            // calls that target the first edit with this guid.
+            var edit = document.Edits.Find(e => e.Guid == stepWrapper.EditGuid);
```
[the remark this answers is not in the document]
Comment (imnasnainaec) [reply Backend/Controllers/UserEditController.cs]: Duplicate guids in one user edit are only reachable through the concurrent first-write race already noted in AddGoalToUserEdit, and the duplicates there are two copies of the same goal, so the worst case is the step landing on the losing copy. Threading StoredEdit.Id through the service and repo API would add 10-15 lines of plumbing. If we want to close it properly, we can make the {projectId, userEditId, guid} index unique in a future pr. (That needs a dedupe pass first, since pre-migration data could already contain duplicates.)

File: database/migrate-useredits-to-edits-collection.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+//   { _id: ObjectId, projectId: string, userEditId: string,
+//     guid: UUID, goalType: int, stepData: [string], changes: string, modified: Date? }
+// Migrated UserEditsCollection `edits` field (must match StoredUserEdit): [ObjectId, ...]
+
+var userEdits = db.getCollection("UserEditsCollection");
+var edits = db.getCollection("EditsCollection");
+
+// Old-format documents have embedded edit objects (each with a `guid` field);
+// migrated documents hold plain ObjectIds, which have no subfields.
+var oldFormat = { "edits.0.guid": { $exists: true } };
```
Comment (jasonleenaylor) [inline database/migrate-useredits-to-edits-collection.js:42]: I might be wrong about this — worth a look rather than a change.

`oldFormat` tests `edits.0` only. Could an `edits` array ever hold both embedded edits and a ref?

The path I think could do it is the window in deploy step 3-4, while the new backend is up on unmigrated data. `UpdateUserEditGoal` doesn't read the UserEdit: `IsViolationEdit` reads the User document, and `AddGoalToUserEdit` is `ReplaceEdit || AddEdit`. `ReplaceEdit` would miss, so `AddEdit` would `$push` an ObjectId onto an array of embedded edits. On the frontend, `asyncAddGoal` gets the userEditId from local storage (`GoalActions.ts:234`), so a goal tile click goes straight to that POST with no read first.

If that can happen, the concern is what this script then does with it: `edits.deleteMany({userEditId})` at line 65 would drop the edit `AddEdit` inserted, `doc.edits.map` would read the ObjectId as embedded and get `undefined` for `guid`, and I don't think any of the verification checks would notice — the ref resolves, projectId matches, nothing is unreferenced or orphaned.

You know the deploy better than I do, so this may be closed off by how it's actually run. If it isn't, asserting every element has a `guid` before the map would turn it into a refusal to migrate.

**Review by jasonleenaylor — COMMENTED** [review]
One question inline on the migration script. I'm unsure about it and it may well be closed off by how the deploy is actually run, so it's a look-into rather than a change request.

_This review was assisted by Claude Opus 5._

**Review by jasonleenaylor — APPROVED** [review]
LGTM

File: database/migrate-useredits-to-edits-collection.js [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+  // A kept ref can share a guid with an embedded edit: advancing a step in an
+  // already-started goal makes ReplaceEdit miss, so the pushed ref holds the newer state
+  // of that same goal. The ref wins, and the stale embedded copy is skipped rather than
+  // left as a duplicate guid for the backend's first-match reads to shadow.
+  var supersededGuids = new Set();
+  keptRefs.forEach(function (ref) {
+    var kept = edits.findOne({ _id: ref }, { guid: 1 });
+    if (kept) {
+      supersededGuids.add(String(kept.guid));
```
[the remark this answers is not in the document]
Comment (imnasnainaec) [reply database/migrate-useredits-to-edits-collection.js]: Done.
